pkg/persistence/rethinkdb/store.go

```diff
@@ -0,0 +1,99 @@
+package rethinkdb
+
+import (
+	"strings"
+	"time"
+
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/etc"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/harbor"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/job"
+	"github.com/aquasecurity/harbor-scanner-trivy/pkg/persistence"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/xerrors"
+	r "gopkg.in/rethinkdb/rethinkdb-go.v6"
+)
+
+type rethinkdbScanJob struct {
+	LastUpdatedAt time.Time `json:"last_updated_at"`
+	job.ScanJob
+}
+
+type store struct {
+	conn r.QueryExecutor
+	cfg  etc.Rethink
+}
+
+func NewStore(conn r.QueryExecutor, cfg etc.Rethink) persistence.Store {
+	return &store{
+		conn: conn,
+		cfg:  cfg,
+	}
+}
+
+func (s *store) Create(scanJob job.ScanJob) error {
+	logrus.WithFields(logrus.Fields{
+		"scan_job_id":     scanJob.ID,
+		"scan_job_status": scanJob.Status.String(),
+		"rethinkdb_table": s.cfg.ScansTable,
+	}).Debug("Saving scan job")
+
+	rsj := rethinkdbScanJob{
+		LastUpdatedAt: time.Now().UTC(),
+		ScanJob:    scanJob,
+	}
+
+	res, err := r.Table(s.cfg.ScansTable).Insert(rsj).RunWrite(s.conn)
+	if err != nil || res.Inserted != 1 {
+		return xerrors.Errorf("inserting scan job failed: %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) Get(scanJobID string) (*job.ScanJob, error) {
+	scanJob := new(job.ScanJob)
+	if err := r.Table(s.cfg.ScansTable).Get(scanJobID).ReadOne(scanJob, s.conn); err != nil {
+		if err == r.ErrEmptyResult {
+			return nil, nil
+		}
+
+		return nil, xerrors.Errorf("getting scan job failed: %w", err)
+	}
+
+	return scanJob, nil
+}
+
+func (s *store) UpdateStatus(scanJobID string, newStatus job.ScanJobStatus, errors ...string) error {
+	logrus.WithFields(logrus.Fields{
+		"scan_job_id": scanJobID,
+		"new_status":  newStatus.String(),
+	}).Debug("Updating status for scan job")
+
+	res, err := r.Table(s.cfg.ScansTable).Get(scanJobID).Update(map[interface{}]interface{}{
+		"status": newStatus,
+		"error":  strings.Join(errors, " --- "),
+		"last_updated_at": time.Now().UTC(),
+	}).RunWrite(s.conn)
+	if err != nil || res.Replaced != 1 {
+		return xerrors.Errorf("updating scan job status failed: %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) UpdateReport(scanJobID string, newReport harbor.ScanReport) error {
+	logrus.WithFields(logrus.Fields{
+		"scan_job_id": scanJobID,
+	}).Debug("Updating report for scan job")
+
+	res, err := r.Table(s.cfg.ScansTable).Get(scanJobID).Update(map[interface{}]interface{}{
```

Comment:
Probably would be good to encapsulate in a data structure of some sort.

Reply:
You mean creating a new private struct and use it instead of the map[interface{}]interface{}?

Reply:
You can probably reuse this struct, https://github.com/hrist0stoichev/harbor-scanner-trivy/blob/RethinkDb/pkg/job/model.go create a struct with the update fields populated (and since it looks properly annotated with `omitempty` it should only update those field that are set in the struct.

Reply:
If I want to reuse the structure that you've linked, I need to add `omitempty` to the ID field as well. I thought a lot about how this should be done and I don't see a clear winner. I've identified 3 possibilities and all of them have some drawbacks.

1. Use the `job.ScanJob` struct and add `omitempty` to the `ID` and `Status` fields. This is the easiest solution but what bugs me is that `omitempty` somehow implies that those fields are not mandatory when, in fact, they are. One would be able to use the `Create` method to insert an empty `scanJob` and this sounds wrong. Maybe some validation can be added to the `Create` method but then this is tied to the particular implementation. Also, if someone deletes the `omitempty` at some point, the code will stop working and there won't be a clear indication for that. Maybe it would be best to add `rethinkdb` annotations to the fields instead of relying on the `json` annotations. Finally, I see a potential issue in updating one field (the status, for example) by passing a struct that contains multiple fields and relying on annotations. Consider the following situation - we want to update the `Report` field and write the following code `r.Table(...).Get(...).Update(ScanJob{ Report: report })`. This will in fact update the `Report`, but it will also update the `Status` because the default value for int is 0 and even though we have `omitempty` on the `Status` field - it will get updated. (maybe the library treats 0 as empty value, I am not sure, but you get my point).
2. Use `map` and only update the fields that are supposed to be updated - after all the method is called `UpdateStatus`. The biggest problem I see with this approach is that if someone changes the name of the `Status` field, the code will start failing during runtime. At least, we have some integration tests for the store.
3. The same as 2 but using a private struct instead of a map. I don't really have a strong preference for which of the two is better. Maybe using a struct has a slight advantage for code readability, etc.

What do you think?

Reply:
For point 1, yes I do believe libraries treat 0 as empty value, but this causes an issue if you actually need a value of 0.  The workaround this for many times for update fields is to have all fields be of a pointer to the actual type so we can nil them (or not initialize them) if not needed.  Yes possibly so we don't muck around with the existing struct, I would think approach 3 could potentially be good using pointer fields as I mentioned with omitempty.